packages/react-client/src/devices/MiddlewareManager.ts

```diff
@@ -0,0 +1,58 @@
+import type { TrackMiddleware } from "../types/public";
+import { setupOnEndedCallback } from "../utils/track";
+
+export type MiddlewareMedia = {
+  stream: MediaStream;
+  track: MediaStreamTrack;
+};
+
+export class MiddlewareManager {
+  private middleware: TrackMiddleware | null = null;
+  private clearMiddlewareFn: (() => void) | null = null;
+
+  public updateMedia(track: MediaStreamTrack | null): MiddlewareMedia | null {
+    return this.applyMiddleware(track, this.middleware);
+  }
+
+  public setTrackMiddleware(
+    middleware: TrackMiddleware | null,
+    rawTrack: MediaStreamTrack | null,
+  ): MiddlewareMedia | null {
+    this.middleware = middleware;
+    return this.applyMiddleware(rawTrack, middleware);
+  }
```

Comment:
Controversial Decision: The `updateMedia` and `setTrackMiddleware` functions mutate the state and return a new object. While this violates the Command Query Separation principle, I believe it's acceptable in this context. If we decide not to return from these functions, we need to store the result of `applyMiddleware` and then expose a method that allows us to read this field.

Reply:
Or just use `updateMedia` explicitly when setting a new middleware. It separates the purposes really well.

Reply:
Doing so `setTrackMiddleware` still violates the Command Query Separation. `applyMiddleware` mutates internal state